File: src/main/scala/numtowords/NumberToWords.scala
package numtowords


object NumberToWords {

  private val smallNumberNameMap = Map(
    0L -> "",
    1L -> "one",
    2L -> "two",
    3L -> "three",
    4L -> "four",
    5L -> "five",
    6L -> "six",
    7L -> "seven",
    8L -> "eight",
    9L -> "nine",
    10L -> "ten",
    11L -> "eleven",
    12L -> "twelve",
    13L -> "thirteen",
    14L -> "fourteen",
    15L -> "fifteen",
    16L -> "sixteen",
    17L -> "seventeen",
    18L -> "eighteen",
    19L -> "nineteen",
    20L -> "twenty",
    30L -> "thirty",
    40L -> "forty",
    50L -> "fifty",
    60L -> "sixty",
    70L -> "seventy",
    80L -> "eighty",
    90L -> "ninety",
    100L -> "hundred"
  )
  
  private val largeNumberNames = List("", "thousand", "million", "billion", "trillion", "quadrillion", "quintillion")
  
  def numberToWords(i: Long): String = i match {
    case _ if i < 0 => "minus " + numberToWords(-i)
    case 0 => "zero"
    case _ => normalizeWhiteSpace(convert(i, largeNumberNames).trim)
  }
  
  private def convert(i: Long, dividers: List[String]): String = {
    if (i > 0) {
      convert(i / 1000, dividers.tail) + " " + convertNamedLargeNumber(i % 1000, dividers.head)
    } else ""
  }
  
  private def convertNamedLargeNumber(i: Long, largeNumberName: String): String = {
    if (i > 0) {
      convertHundreds(i / 100) + " " +  convertTens(i % 100) + " " + largeNumberName 
    } else ""
  }
  
  private def convertHundreds(i: Long): String = {
    if (i > 0) {
      smallNumberNameMap(i) + " " + smallNumberNameMap(100L)
    } else {
       ""
    }
  }
  
  private def convertTens(i: Long): String = {
    if (i <= 20) {
      smallNumberNameMap(i)
    } else {
      smallNumberNameMap(i - i % 10) + " " + smallNumberNameMap(i % 10)
    }
  }  
  
  private def normalizeWhiteSpace(s: String) = {
    s replaceAll("\\s+", " ")
  }
}

File: src/main/java/j_numtowords/NumberToWords.java
package j_numtowords;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class NumberToWords {

	private static final  Map<Long, String> smallNumberNameMap = new HashMap<Long, String>();
	static {
		smallNumberNameMap.put(0L, "");
		smallNumberNameMap.put(1L, "one");
		smallNumberNameMap.put(2L, "two");
		smallNumberNameMap.put(3L, "three");
		smallNumberNameMap.put(4L, "four");
		smallNumberNameMap.put(5L, "five");
		smallNumberNameMap.put(6L, "six");
		smallNumberNameMap.put(7L, "seven");
		smallNumberNameMap.put(8L, "eight");
		smallNumberNameMap.put(9L, "nine");
		smallNumberNameMap.put(10L, "ten");
		smallNumberNameMap.put(11L, "eleven");
		smallNumberNameMap.put(12L, "twelve");
		smallNumberNameMap.put(13L, "thirteen");
		smallNumberNameMap.put(14L, "fourteen");
		smallNumberNameMap.put(15L, "fifteen");
		smallNumberNameMap.put(16L, "sixteen");
		smallNumberNameMap.put(17L, "seventeen");
		smallNumberNameMap.put(18L, "eighteen");
		smallNumberNameMap.put(19L, "nineteen");
		smallNumberNameMap.put(20L, "twenty");
		smallNumberNameMap.put(30L, "thirty");
		smallNumberNameMap.put(40L, "forty");
		smallNumberNameMap.put(50L, "fifty");
		smallNumberNameMap.put(60L, "sixty");
		smallNumberNameMap.put(70L, "seventy");
		smallNumberNameMap.put(80L, "eighty");
		smallNumberNameMap.put(90L, "ninety");
		smallNumberNameMap.put(100L, "hundred");
	}
	
	private static final String[] largeNumberNames = { "", "thousand", "million", "billion", "trillion", "quadrillion", "quintillion" };
	
	public static String numberToWords(final long i) {
		return i < 0?
				"minus " + numberToWords(-i) :
					i == 0?
							"zero" :
								normaliseWhiteSpace(convert(i, 0));
	}

	private static String convert(final Long i, final int largeNumberNamesIndex) {
	    if (i > 0) {
	        return convert(i / 1000, largeNumberNamesIndex + 1) + " " + convertNamedLargeNumber(i % 1000, largeNumberNames[largeNumberNamesIndex]);
	    } else {
	    	return "";		
	    }
	}
	
	private static String convertNamedLargeNumber(final long i, final String largeNumberName) {
	    if (i > 0) {
	        return convertHundreds(i / 100) + " " +  convertTens(i % 100) + " " + largeNumberName;
	    } else {
	    	return "";
	    }
	}
	
	private static String convertHundreds(final Long i) {
		if (i > 0) {
			return smallNumberNameMap.get(i) + " " + smallNumberNameMap.get(100L);
		} else {
			return "";			
		}
	}
	
	private static String convertTens(final Long i) {
		if (i <= 20) {
			return smallNumberNameMap.get(i);
		} else {
			final long units = i % 10;
			return (smallNumberNameMap.get(i - units) + " " + smallNumberNameMap.get(units)).trim();			
		}
	}
	
	private static String normaliseWhiteSpace(final String s) {
		return s.trim().replaceAll("\\s+", " ");
	}
}
